src/music/dmusic.cpp

```diff
@@ -19,86 +19,1093 @@
+								case MIDIST_SONGSEL: /* song select */
+									data++;
+									remaining--;
+									break;
+								case MIDIST_SONGPOSPTR: /* song position pointer */
+									data += 2;
+									remaining -= 2;
+									break;
+								default: /* remaining have no data bytes */
+									break;
+							}
+							break;
+					}
+				}
+
+				current_block++;
+			}
+
+			/* Anything in the playback buffer? Send it down the port. */
+			DWORD used_buffer = 0;
+			_buffer->GetUsedBytes(&used_buffer);
+			if (used_buffer > 0) {
+				_port->PlayBuffer(_buffer);
+				_buffer->Flush();
+			}
+
+			/* end? */
+			if (current_block == current_file.blocks.size()) {
+				if (current_segment.loop) {
+					current_block = 0;
+					clock->GetTime(&playback_start_time);
+				} else {
+					_playback.do_stop = true;
+				}
+				next_timeout = 0;
+			}
+		}
+	}
+
+	DEBUG(driver, 2, "DMusic: Exiting playback thread");
+
+	/* Turn all notes off and wait a bit to allow the messages to be handled by real hardware. */
+	clock->GetTime(&cur_time);
+	TransmitNotesOff(_buffer, block_time, cur_time);
+	Sleep(_playback.preload_time * 4);
+
+	clock->Release();
+}
+
+static void * DownloadArticulationData(int base_offset, void *data, const std::vector<CONNECTION> &artic)
+{
+	DMUS_ARTICULATION2 *art = (DMUS_ARTICULATION2 *)data;
+	art->ulArtIdx = base_offset + 1;
+	art->ulFirstExtCkIdx = 0;
+	art->ulNextArtIdx = 0;
+
+	CONNECTIONLIST *con_list = (CONNECTIONLIST *)(art + 1);
+	con_list->cbSize = sizeof(CONNECTIONLIST);
+	con_list->cConnections = (ULONG)artic.size();
+	MemCpyT((CONNECTION *)(con_list + 1), &artic.front(), artic.size());
+
+	return (CONNECTION *)(con_list + 1) + artic.size();
+}
+
+static const char *LoadDefaultDLSFile(const char *user_dls)
+{
+	DMUS_PORTCAPS caps;
+	MemSetT(&caps, 0);
+	caps.dwSize = sizeof(DMUS_PORTCAPS);
+	_port->GetCaps(&caps);
+
+	if ((caps.dwFlags & (DMUS_PC_DLS | DMUS_PC_DLS2)) != 0 && (caps.dwFlags & DMUS_PC_GMINHARDWARE) == 0) {
+		DLSFile dls_file;
+
+		if (user_dls == NULL) {
+			/* Try loading the default GM DLS file stored in the registry. */
+			HKEY hkDM;
+			if (SUCCEEDED(RegOpenKeyEx(HKEY_LOCAL_MACHINE, _T("Software\\Microsoft\\DirectMusic"), 0, KEY_READ, &hkDM))) {
+				TCHAR dls_path[MAX_PATH];
+				DWORD buf_size = sizeof(dls_path); // Buffer size as to be given in bytes!
+				if (SUCCEEDED(RegQueryValueEx(hkDM, _T("GMFilePath"), NULL, NULL, (LPBYTE)dls_path, &buf_size))) {
+					TCHAR expand_path[MAX_PATH * 2];
+					ExpandEnvironmentStrings(dls_path, expand_path, lengthof(expand_path));
+					if (!dls_file.LoadFile(expand_path)) DEBUG(driver, 1, "Failed to load default GM DLS file from registry");
+				}
+				RegCloseKey(hkDM);
+			}
+
+			/* If we couldn't load the file from the registry, try again at the default install path of the GM DLS file. */
+			if (dls_file.instruments.size() == 0) {
+				static const TCHAR *DLS_GM_FILE = _T("%windir%\\drivers\\gm.dls");
```

Comment:
This is wrong in both Windows 10 and Windows Me, and presumably also everything in between. `%WINDIR%\System32\drivers\GM.DLS` should be correct, also on 64 bit systems.

Reply:
Indeed, I think I had that right and then made some last-minute changes.